src/rnnlm/rnnlm-simple-looped.cc

```diff
@@ -0,0 +1,200 @@
+  info_(other.info_), computer_(other.computer_), feats_(other.feats_),
+  current_nnet_output_(other.current_nnet_output_),
+  current_log_post_offset_(other.current_log_post_offset_)
+{}
+
+void RnnlmSimpleLooped::TakeFeatures(
+    const std::vector<int32> &word_indexes) {
+  KALDI_ASSERT(word_indexes.size() == feats_.NumRows());
+  std::vector<std::vector<std::pair<MatrixIndexT, BaseFloat> > >
+      pairs(word_indexes.size());
+  for (int32 i = 0; i < word_indexes.size(); i++) {
+    std::pair<MatrixIndexT, BaseFloat> one_hot_index(word_indexes[i], 1.0);
+    std::vector<std::pair<MatrixIndexT, BaseFloat> > row(1, one_hot_index);
+    pairs[i] = row;
+  }
+  SparseMatrix<BaseFloat> feats_temp(feats_.NumCols(), pairs);
+  feats_.Swap(&feats_temp);
+  // resets offset so that AdvanceChunk() would be called in GetOutput() and
+  // GetNnetOutputForFrame() after taking new features
+  current_log_post_offset_ = -1;
+}
+
+//void RnnlmSimpleLooped::GetNnetOutputForFrame(
+//    int32 frame, VectorBase<BaseFloat> *output) {
+//  KALDI_ASSERT(frame >= 0 && frame < feats_.NumRows());
+//  if (frame >= current_log_post_offset_ + current_nnet_output_.NumRows())
+//    AdvanceChunk();
+//  output->CopyFromVec(current_nnet_output_.Row(frame -
+//                                               current_log_post_offset_));
+//}
+
+BaseFloat RnnlmSimpleLooped::LogProbOfWord(int32 word_index,
+                               const CuVectorBase<BaseFloat> &hidden) const {
+  const CuMatrix<BaseFloat> &word_embedding_mat = info_.word_embedding_mat;
+  BaseFloat log_prob;
+
+  if (info_.opts.force_normalize) {
+    CuVector<BaseFloat> log_probs(word_embedding_mat.NumRows());
+
+    log_probs.AddMatVec(1.0, word_embedding_mat, kTrans, hidden, 0.0);
+    log_probs.ApplySoftMax();
+    log_probs.ApplyLog();
+    log_prob = log_probs(word_index);
+  } else {
+    log_prob = VecVec(hidden, word_embedding_mat.Row(word_index));
+  }
+  return log_prob;
+}
+
+CuVector<BaseFloat>* RnnlmSimpleLooped::GetOutput(int32 frame) {
+  KALDI_ASSERT(frame >= 0 && frame < feats_.NumRows());
+  if (frame >= current_log_post_offset_ + current_nnet_output_.NumRows())
+    AdvanceChunk();
+
+//  int32 embedding_dim = info_.word_embedding_mat.NumCols();
+//  int32 num_words = info_.word_embedding_mat.NumRows();
+
+
+  CuMatrix<BaseFloat> current_nnet_output_gpu;
+  current_nnet_output_gpu.Swap(&current_nnet_output_);
+  const CuSubVector<BaseFloat> hidden(current_nnet_output_gpu,
+                                      frame - current_log_post_offset_);
+  return new CuVector<BaseFloat>(hidden);
+//
+//  // swap the pointer back so that this function can be called multiple times
+//  // with the same returned value before taking next new feats
+//  current_nnet_output_.Swap(&current_nnet_output_gpu);
+//  return log_prob;
+}
+
+void RnnlmSimpleLooped::AdvanceChunk() {
+  int32 begin_input_frame, end_input_frame;
+  begin_input_frame = -info_.frames_left_context;
```

Comment:
the frames_left_context and frames_rigth_context will be zero, this code can be simplified a lot.